From: Lucas Stach <l.stach@pengutronix.de>
Date: Thu, 22 Feb 2018 11:54:21 +0100
Subject: etnaviv: GC7000: flush TX descriptor and instruction cache

The etnaviv kernel driver will only ever flush write caches. As both
the TX descriptor and instruction cache are read caches they must be
flushed from the user cmdstream at an appropriate time.

Signed-off-by: Lucas Stach <l.stach@pengutronix.de>
---
 src/gallium/drivers/etnaviv/etnaviv_context.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/gallium/drivers/etnaviv/etnaviv_context.c b/src/gallium/drivers/etnaviv/etnaviv_context.c
index 2d7da98..963754e 100644
--- a/src/gallium/drivers/etnaviv/etnaviv_context.c
+++ b/src/gallium/drivers/etnaviv/etnaviv_context.c
@@ -390,9 +390,16 @@ etna_cmd_stream_reset_notify(struct etna_cmd_stream *stream, void *priv)
        * before command stream submission. It does not need flushing if the
        * referenced image data changes.
        */
+      etna_set_state(stream, VIVS_NTE_DESCRIPTOR_FLUSH, 0);
       etna_set_state(stream, VIVS_GL_FLUSH_CACHE,
             VIVS_GL_FLUSH_CACHE_DESCRIPTOR_UNK12 |
             VIVS_GL_FLUSH_CACHE_DESCRIPTOR_UNK13);
+
+      /* Icache invalidate (should do this on shader change?) */
+      etna_set_state(stream, VIVS_VS_ICACHE_INVALIDATE,
+            VIVS_VS_ICACHE_INVALIDATE_UNK0 | VIVS_VS_ICACHE_INVALIDATE_UNK1 |
+            VIVS_VS_ICACHE_INVALIDATE_UNK2 | VIVS_VS_ICACHE_INVALIDATE_UNK3 |
+            VIVS_VS_ICACHE_INVALIDATE_UNK4);
    }
 
    ctx->dirty = ~0L;
